Rainbow highlighting spike 🌈

killercup opened the pull request:
Very simple approach: For each identifier, set the hash of the range
where it's defined as its 'id' and use it in the VSCode extension to
generate unique colors.

Thus, the generated colors are per-file. They are also quite fragile,
and I'm not entirely sure why. Looks like we need to make sure the
same ranges aren't overwritten by a later request?

killercup commented on crates/ra_ide_api/src/syntax_highlighting.rs:
```diff
@@ -32,59 +33,72 @@ pub(crate) fn highlight(db: &RootDatabase, file_id: FileId) -> Vec<HighlightedRa
-            k if is_control_keyword(k) => "keyword.control",
-            k if k.is_keyword() => "keyword",
+            NAME => if let Some(name) = node.as_ast_node::<ast::Name>() {
+                ("variable", Some(hash(name.syntax().range())))
```
This is certainly not ideal. I just wanted it to show something for now.

killercup commented on editors/code/src/highlighting.ts:
```diff
@@ -84,17 +101,32 @@ export class Highlighter {
-                .get(d.tag)!
-                .push(Server.client.protocol2CodeConverter.asRange(d.range));
+            
+            if (d.id) {
```
To make this opt-in, add checking a settings flag here

matklad replied:
:taco: 

EDIT: that should be :+1:, github autocomplete is ... too smart?

killercup commented on editors/code/src/highlighting.ts:
```diff
@@ -6,6 +7,20 @@ import { Server } from './server';
+
+    const h = randomInt(0, 360);
+    const s = randomInt(42, 98);
+    const l = shade === 'light' ? randomInt(15, 40) : randomInt(40, 90);
```
Just a random guess

matklad commented on crates/ra_ide_api/src/syntax_highlighting.rs:
```diff
@@ -32,59 +33,74 @@ pub(crate) fn highlight(db: &RootDatabase, file_id: FileId) -> Vec<HighlightedRa
+                        Some(Def(ModuleDef::Trait(_))) => ("type", None),
+                        Some(Def(ModuleDef::TypeAlias(_))) => ("type", None),
+                        Some(SelfType(_)) => ("type", None),
+                        Some(Pat(ptr)) => ("variable", Some(hash(ptr.syntax_node_ptr().range()))),
```
Hashing position is not ideal: the colors would change after every modification.

I think we should hash the name of variable. Note that this will cause all shadowed versions of var to be colored the same, but I think that's ok as a protopyte. To deal with shadowing, we can hash a tuple of `(name, N)`, where N means that this Nth definition of name in scope (so, `(foo, 0)` for first one, `(foo, 1)` for first shadowed, etc)

matklad commented on crates/ra_ide_api/src/snapshots/tests__highlighting.snap:
```diff
@@ -8,139 +8,185 @@ Ok(
             range: [25; 31),
-            tag: "keyword"
+            tag: "keyword",
+            id: None
```
You'll need to re-accept the tests with fresh rust 1.35

matklad commented on crates/ra_ide_api/src/syntax_highlighting.rs:
```diff
@@ -32,59 +33,74 @@ pub(crate) fn highlight(db: &RootDatabase, file_id: FileId) -> Vec<HighlightedRa
-            k if is_control_keyword(k) => "keyword.control",
-            k if k.is_keyword() => "keyword",
+            NAME => {
+                if let Some(name) = node.as_ast_node::<ast::Name>() {
```
I think we should check if the parent of this name is `PatBinding`, an only then use `variable`, falling back to `function` overwise

matklad commented on crates/ra_ide_api/src/syntax_highlighting.rs:
```diff
@@ -32,59 +33,74 @@ pub(crate) fn highlight(db: &RootDatabase, file_id: FileId) -> Vec<HighlightedRa
+            TYPE_ALIAS_DEF | TYPE_ARG | TYPE_PARAM => ("type", None),
+            INT_NUMBER | FLOAT_NUMBER | CHAR | BYTE => ("literal", None),
+            LIFETIME => ("parameter", None),
+            T![unsafe] => ("keyword.unsafe", None),
```
I wonder if we can special case variables somehow, instead of adding `None` everywhere.

Perhaps just an `let mut id = None` ?

matklad commented on crates/ra_lsp_server/src/req.rs:
```diff
@@ -129,6 +129,7 @@ pub struct PublishDecorationsParams {
 pub struct Decoration {
     pub range: Range,
     pub tag: &'static str,
+    pub id: Option<String>,
```
`id` seems like an unfortunate name.

Let's make it `binding_hash: Option<u64>`, with a short comment explaining how we use this for highlighting

matklad commented on crates/ra_syntax/src/syntax_node.rs:
```diff
@@ -523,6 +523,10 @@ impl<'a> SyntaxElement<'a> {
         }
     }
 
+    pub fn as_ast_node<T: AstNode>(&self) -> Option<&T> {
```
i'd rather do `it.as_node().and_then(T::Cast)` on the call-site. I am somehow not a fan of a lot of small helper methods, because the make call-site harder to understand (b/c you need to goto defenition to the helper method to understand what's going on).

matklad commented:
For the reference, here's the relevant code from IntelliJ Rust (yes, my life is a groundhog day movie): https://github.com/intellij-rust/intellij-rust/blob/037f43bfd76785dad8aac989e2fd1f2118f212b9/src/main/kotlin/org/rust/ide/highlight/RsRainbowVisitor.kt#L15-L45

RainbowHighlighting is a much better name for this feature....

matklad commented:
https://github.com/rust-analyzer/rust-analyzer/pull/1327 adds ability to hightlight rust code as HTML. Also, turns out that `!` is a valid character in git branch name, I should use it more

killercup commented:
Rebased on top of html output, and added `--rainbow` to CLI. It now uses the name and a shadowing counter. This of course only works for local variables, which is all that is supported for now. Function/struct/enum names, as well as struct field names and enum variants can be supported if we can get a good unique ID for them.

jens1o commented:
I wonder whether this could be made an optional feature, as personally I'd never use it and it increases compile-time (again) by requiring `rand` with all of its dependency-tree.

killercup commented:
Runtime-optional, sure. A compile-time feature doesn't seem worth it
though; random was also already in the lock file.

On Sun, 26 May 2019, 08:22 Jens Hausdorf, <notifications@github.com> wrote:

> I wonder whether this could be made an optional feature, as personally I'd
> never use it and it increases compile-time (again) by requiring rand with
> all of its dependency-tree.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-analyzer/rust-analyzer/pull/1319?email_source=notifications&email_token=AAAE4X3J7IJX3ID33PX64NTPXIUA3A5CNFSM4HPJ3HCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWH7GCY#issuecomment-495973131>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAE4XZCQUZOD3LP6GVIWN3PXIUA3ANCNFSM4HPJ3HCA>
> .
>

matklad commented:
>by requiring rand with all of its dependency-tree.

Unfortunately, we already depend on `rand` :-(

And yeah, one day someone (probably me) would have to look into seriously cutting the number of dependencies and optimizing compile times

matklad commented:
r=me with config option added (don't forget to add the docs for it as well, to the `docs/user/features.md`)

bors delegate+

jens1o commented:
> Unfortunately, we already depend on rand :-(

Ah, pardon me. I didn't saw that.

killercup commented:
bors r+

✨ 🌈 ✨
